src/base/iter.rs

```diff
@@ -396,6 +418,96 @@ impl<'a, T: Scalar, R: Dim, C: Dim, S: 'a + RawStorageMut<T, R, C>> ExactSizeIte
 {
     #[inline]
     fn len(&self) -> usize {
-        self.ncols() - self.curr
+        self.range.len()
+    }
+}
+
+impl<'a, T: Scalar, R: Dim, C: Dim, S: 'a + RawStorageMut<T, R, C>> DoubleEndedIterator
+    for ColumnIterMut<'a, T, R, C, S>
+{
+    fn next_back(&mut self) -> Option<Self::Item> {
+        debug_assert!(self.range.start <= self.range.end);
+        if !self.range.is_empty() {
+            self.range.end -= 1;
+            debug_assert!(self.range.end < self.ncols());
+            debug_assert!(self.range.end >= self.range.start);
+            Some(unsafe { (*self.mat).column_mut(self.range.end) })
+        } else {
+            None
+        }
+    }
+}
+
+/// implementations for parallel iteration with rayon
+#[cfg(feature = "rayon")]
+mod parallel {
+    use super::*;
+    use rayon::iter::plumbing::Producer;
+
+    impl<'a, T, R: Dim, Cols: Dim, S: RawStorage<T, R, Cols>> Producer for ColumnIter<'a, T, R, Cols, S>
+    where
+        T: Send + Sync + Debug + PartialEq + Clone + 'static,
+        S: Sync,
+    {
+        type Item = MatrixSlice<'a, T, R, U1, S::RStride, S::CStride>;
+        type IntoIter = ColumnIter<'a, T, R, Cols, S>;
+
+        fn split_at(self, index: usize) -> (Self, Self) {
+            // the index is relative to the size of this current iterator
+            // it will always start at zero so it serves as an offset
+            let left = Self {
+                mat: self.mat,
+                range: self.range.start..(self.range.start + index),
+            };
+
+            let right = Self {
+                mat: self.mat,
+                range: (self.range.start + index)..self.range.end,
+            };
+            (left, right)
+        }
+
+        fn into_iter(self) -> Self::IntoIter {
+            self
+        }
+    }
+
+    impl<'a, T: Scalar, R: Dim, C: Dim, S: 'a + RawStorageMut<T, R, C>> Producer
```

Comment:
Note: The fact that `Producer` is implemented for `ColumnIter`/`ColumnIterMut` will be part of the public API. I'm not sure if this is desirable or not. It seems to me that this ought to be only an implementation detail used in the `ParallelIndexedIterator` impl. Something to think about anyway.

Reply:
I'm in two minds about this. The other way I can see to solve this is to create basically a 1-to-1 code duplicate of the iterators that are not part of the public API. I don't like code duplication, but I can also see how you might want to hide the implementation. I'll go whichever way you like and maybe there is a more elegant solution that I am not seeing right now.

Reply:
I think you could create a private wrapper type around the iterator and implement the `Producer` for that. That should result in fairly minimal code duplication I think?

Reply:
Great idea. You mean like a newtype, right?

Reply:
That's right!

Reply:
Okay, I refactored this into a separate producer type that is not publicly visible. Let me know if you like that implementation and thanks for all the improvements thus far.